### PR TITLE
Optimize "ORDER BY + LIMIT" queries for speed / memory with special TopK operator

### LINKED ISSUE #7196: Optimize "LIMIT" queries for speed / memory with special TopK operator
### Is your feature request related to a problem or challenge?

This pattern is common:

```
SELECT c1, c2
FROM t
ORDER BY c3
LIMIT 10
```

For example we have queries in IOx like the following (this is the same pattern @NGA-TRAN describes on https://github.com/apache/arrow-datafusion/issues/7162)

```
SELECT tag, value1, ...
FROM t
WHERE other_column = 'foo'
ORDER BY time
LIMIT 10
```

### Describe the solution you'd like

If the data *IS NOT* already sorted, what happens today is a plan like

```
LIMIT(fetch=10)
  SORT(sort_exprs=[c3] fetch=10)
    SCAN(...)
```

And the Sort can take partial advantage of the fetch -- and it will be better after @gruuya 's change in https://github.com/apache/arrow-datafusion/pull/7180

We can probably do better still with a special operator like the following that uses some specialized structure (perhaps some type of heap)

```
TOPK(fetch=10, sort_exprs=[c3])
    SCAN(...)
```

### Describe alternatives you've considered

If the data is already sorted the right way, DataFusion can just read first N values and stop as described on https://github.com/apache/arrow-datafusion/issues/7162

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Dandandan**
## Which issue does this PR close?

Closes #7196
Closes https://github.com/apache/arrow-datafusion/pull/7250

## Rationale for this change

Adding TopK to DataFusion limits the resources (memory, CPU) needed for `SELECT .. ORDER BY [..] LIMIT N` type of queries.
@alamb implemented most of the changes necessary, this PR does some final update / clean up of the code.

### CODE REVIEW

**Comment by alamb** [comment]
Thank you @Dandandan  -- let me know if you would like help finishing up this PR. It has been on my list but I haven't had a chance yet.

Maybe I could make a PR that changed the display of plans to show when topk was being used 🤔

File: datafusion/physical-plan/src/topk/mod.rs
Comment (alamb) [inline datafusion/physical-plan/src/topk/mod.rs:672]: BTW I think @tustvold  basically did this upstream in arrow 47.0.0 but now I am struggling to find the PR.

So in other words I think we can remove the `interleave_and_repack_dictionary` code. I can test with my high cardinality dictionary usecase if that would help

Comment (Dandandan) [reply datafusion/physical-plan/src/topk/mod.rs:672]: Thanks - I made the change

File: datafusion/physical-plan/src/topk/mod.rs
Comment (Dandandan) [inline datafusion/physical-plan/src/topk/mod.rs:163]: Is this TODO still applicable @alamb? I don't think we can do either one of them as we need to visit all input rows and compare them with the max?

One possibility I'm seeing is to filter out rows early (per batch) that are bigger than the current max (which might be better vectorizable than the per-row logic).

File: datafusion/physical-plan/src/sorts/sort.rs
Comment (Dandandan) [inline datafusion/physical-plan/src/sorts/sort.rs:771]: I'm not sure if we would like to do this? I think there are some other `ExecutionPlan` nodes that have the algorithm depend on one of the parameters (for example: HashAggregate `mode`s) .

Comment (alamb) [reply datafusion/physical-plan/src/sorts/sort.rs:771]: I think in general it would be good to be able to tell what operator was going to be used from looking at the plan. However, I think we can do so as a follow on PR -- I can file a ticket.

File: datafusion/sqllogictest/test_files/aal.slt
Comment (alamb) [inline datafusion/sqllogictest/test_files/aal.slt:1]: this file probably at least needs to be renamed to topk.slt or something other than my initials (`AAL` 😆 )

I am also not sure it is the best test of the topk coverage -- I had this for quickly iterating during development

File: datafusion/sqllogictest/test_files/window.slt
Comment (alamb) [inline datafusion/sqllogictest/test_files/window.slt:2679]: added ts to show that the first two values are tied, and that the output is correct ✅

File: datafusion/physical-plan/src/topk/mod.rs
Comment (alamb) [inline datafusion/physical-plan/src/topk/mod.rs:70]:
```suggestion
/// K=3 elements, reducing the total amount of required buffer memory.
```

File: datafusion/physical-plan/src/topk/mod.rs
Comment (alamb) [inline datafusion/physical-plan/src/topk/mod.rs:125]:
```suggestion
        // TODO there is potential to add special cases for single column sort fields
        // to improve performance
        let row_converter = RowConverter::new(sort_fields)?;
```

File: datafusion/physical-plan/src/topk/mod.rs
Comment (alamb) [inline datafusion/physical-plan/src/topk/mod.rs:143]:
```suggestion
    /// Insert `batch`, remembering if any of its values are among
```

File: datafusion/physical-plan/src/topk/mod.rs
Comment (alamb) [reply datafusion/physical-plan/src/topk/mod.rs:163]: I like the suggestion to try and filter earlier. However, I think it should be driven by profiling and benchmarks. Maybe put that idea in comments as a "Potential future improvement" ?

File: datafusion/physical-plan/src/topk/mod.rs
Comment (alamb) [inline datafusion/physical-plan/src/topk/mod.rs:262]: I think this comment is out of date (and mostly due to my misunderstanding of how the BinaryHeap worked 😆 )

```suggestion
```

File: datafusion/physical-plan/src/topk/mod.rs
Comment (alamb) [inline datafusion/physical-plan/src/topk/mod.rs:381]: This is like `interleave` but for `RecordBatches`. @wjones127  was asking about something similar for `take` for `RecordBatches` as well. Maybe it would be a helpful API to add 🤔

File: datafusion/physical-plan/src/topk/mod.rs
Comment (alamb) [inline datafusion/physical-plan/src/topk/mod.rs:407]: @Dandandan  did you review this heuristic -- I remember I tried it on our high cardinality tracing usecase and it seemed to work well (basically it is needed for large N with random-ish inputs)

**Review by alamb — COMMENTED** [review]
Thank you for picking this up @Dandandan  -- this looks really nice.

I left some small comments / suggestions. Before merging this PR I think we should review the existing LIMIT test coverage. The Minimum coverage I think would is needed needed:
1. multiple-record batch input
2. single and multi-column input,
3. "large N" where N is greater than 20 on randomized input (to ensure the RecordBatch store is covered)




It would also be awesome to do the following (which I can help with / do perhaps):
1. Implement a limit "fuzz" test to check the boundary conditions in a wider range
2. File a follow on ticket to display which algorithm is used in what operator in the explain plan

File: datafusion/sqllogictest/test_files/aal.slt
Comment (Dandandan) [reply datafusion/sqllogictest/test_files/aal.slt:1]: Aha I was wondering what `aal` was standing for 🤣

File: datafusion/physical-plan/src/topk/mod.rs
Comment (Dandandan) [reply datafusion/physical-plan/src/topk/mod.rs:409]: I'll take a look :)

Comment (Dandandan) [reply datafusion/physical-plan/src/topk/mod.rs:409]: I think the heuristic is fine:
- it assures we do compaction at most every `n` (> 20) batches of input or more if batches are utlized
- compaction reduces number of rows to `k`. 20 * 8192 = 163840 rows . If we have some wider columns of 1KB each, the memory usage could be ~200MB with some overhead. Thinking about it, I wonder if we need to trigger the compaction as well if it exceeds the configured memory limit 🤔 
- for very large `k` (a number of times the batch size) we avoid doing compaction too often

We can tweak the heuristic later if there is some cases benefiting from that.

**Review by alamb — COMMENTED** [review]
Thanks @Dandandan  -- I (biasedly) think this PR is looking quite good 👍 

I want to run one final set of performance tests prior to merging (which I am starting now) as well as filing the follow on tickets

**Review by alamb — APPROVED** [review]
Thank you very much @Dandandan .

I think this is ready to to go now. 

I tested with this dataset: [traces.zip](https://drive.google.com/file/d/1SlBGlTrYgcZTITDX_VRG22QZ7IyPX_JR/view?usp=drive_link) (240MB):

| Dataset | Description | Creation |
|--------|--------|--------|
| `traces` | directory of parquet files | N/A |
| `traces.parquet | same data as `traces` in a single parquet file | `copy (select * from 'traces') to 'traces.parquet' |
| `traces_ob_random.parquet` | same as `traces.parquet` but `ORDER BY RANDOM` (so topk gets updated a lot)  | `copy (select * from 'traces' ORDER BY random()) to 'traces_oby_random.parquet'` |


| Branch | Query | Time |
|--------|--------|--------|
| `topk` | `select * from 'traces.parquet' order by time desc limit 10` | 2.416 seconds. |
| `main` | " | 3.403 seconds. |
| `topk` | `select * from 'traces.parquet' order by time desc limit 10000` | 3.073 seconds. |
| `main` | " | 3.868 seconds. |
| `topk` | `select * from 'traces_oby_random.parquet' order by time desc limit 10` | 2.403 seconds. |
| `main` | " | 3.500 seconds. |
| `topk` | `select * from 'traces_oby_random.parquet' order by time desc limit 10000` | 4.024 seconds. |
| `main` | " | 3.997 seconds. |
| `topk` | `select * from 'traces order' by time desc limit 10` | 0.750 seconds. |
| `main` | " | 0.902 seconds. |
| `topk` | `select * from 'traces order' by time desc limit 10000` | 2.256 seconds. |
| `main` | " | 1.244 seconds. |


The only query it gets slower for is  large N with multiple files. I believe this is because reconstructing the 10,000 row outputs for each of the partitions, merging them, and then reconstructing the heap is fairly expensive. It would be better in this case to avoid the sort and doing a final topK 

```
TableScan: traces projection=[attributes, duration_nano, end_time_unix_nano, service.name, span.kind, span.name, span_id, time, trace_id, otel.status_code, parent_span_id]                                    GlobalLimitExec: skip=0, fetch=10000                                                                                                                                                                           SortPreservingMergeExec: [time@7 DESC], fetch=10000
      SortExec: fetch=10000, expr=[time@7 DESC]
        ParquetExec: file_groups={16 groups: [...]}
```

I plan to file a follow on ticket for this shortly

**Comment by alamb** [comment]
FYI @gruuya -- it is finally happening

**Comment by Dandandan** [comment]
Thanks @alamb for the review - I plan to test it at our side (Coralogix) and see if there's some follow-up necessary.

> The only query it gets slower for is large N with multiple files. I believe this is because reconstructing the 10,000 row outputs for each of the partitions, merging them, and then reconstructing the heap is fairly expensive. It would be better in this case to avoid the sort and doing a final topK.

This sounds like a good idea, although probably in for distributed usage (e.g. Coralogix) might not be beneficial as we'll need to fetch all partitions instead of doing TopK + merge in a distributed manner.

**Comment by alamb** [comment]
Filed 
* https://github.com/apache/arrow-datafusion/issues/7749 to track adding fuzz tests
* https://github.com/apache/arrow-datafusion/issues/7750 to track making it clearer when TopK is used in the plans

**Pull request merged**
